presto-main/src/main/java/com/facebook/presto/operator/exchange/LocalExchangeSource.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.exchange;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class LocalExchangeSource
+{
+    private static final SettableFuture<?> NOT_EMPTY;
+    static {
+        NOT_EMPTY = SettableFuture.create();
+        NOT_EMPTY.set(null);
+    }
+
+    private final List<Type> types;
+    private final Consumer<LocalExchangeSource> onFinish;
+
+    private final BlockingQueue<PageReference> buffer = new LinkedBlockingDeque<>();
+    private final AtomicLong bufferedBytes = new AtomicLong();
+
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    private SettableFuture<?> notEmptyFuture = NOT_EMPTY;
+
+    @GuardedBy("lock")
+    private boolean finishing;
+
+    public LocalExchangeSource(List<? extends Type> types, Consumer<LocalExchangeSource> onFinish)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.onFinish = requireNonNull(onFinish, "onFinish is null");
+    }
+
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    public LocalExchangeBufferInfo getBufferInfo()
+    {
+        // This must be lock free to assure task info creation is fast
+        // Note: the stats my be internally inconsistent
+        return new LocalExchangeBufferInfo(bufferedBytes.get(), buffer.size());
+    }
+
+    void addPage(PageReference pageReference)
+    {
+        checkNotHoldsLock();
+
+        boolean added = false;
+        SettableFuture<?> notEmptyFuture;
+        synchronized (lock) {
+            // ignore pages after finish
+            if (!finishing) {
```

Comment:
With limit queries, the drivers reading from the exchange complete first.  The writers get notified of this asynchronously, so they maybe in the middle of adding a page when this happens.